app/controllers/CommonFunctionality.scala

```diff
@@ -0,0 +1,138 @@
+ *   * "search" (optional)
+ *   * "filter" (optional)
+ *   * "submit" (mandatory)
+ * - The submit button will indicate what input the user has given:
+ *   * "continue": the user intends to continue the journey onto the next page
+ *   * "filter": the user wishes to apply filters to the current view according to their form input
+ *   * "clear": the user wishes to clear all current filters
+ *   * "pagination_[x]": the user wishes to view page number 'x' of the view.
+ * - When rendering, the search term is not passed explicitly but is stored in the CLIENT_SEARCH_INPUT session value
+ * - When rendering, the filter term is not passed explicitly but is stored in the CLIENT_FILTER_INPUT session value
+ */
+abstract class POSTPaginationHandler[FormData](sessionCacheService: SessionCacheService) {
+  /** A blank form of the type used by this page. */
+  val emptyForm: Form[FormData]
+  /** A function to render the page given a (potentially) pre-filled form. */
+  val renderPage: Form[FormData] => Future[Result]
+  /** HTTP call used by the page to reload itself, if it should need to. The parameters are page number, search term, filter term */
+  val reloadCall: (Option[Int], Option[String], Option[String]) => Call
+  /** A function that will be called when submitting with a valid form, regardless of which button the user has submitted with.
+   *  Typically used to update the session state in some way. */
+  val onSubmit: FormData => Future[Unit]
+  /** The function that will be called when submitting with a "continue" button, and the data is valid. */
+  val onContinue: FormData => Future[Result]
+
+  val checkSubmitTypeForm: Form[String] = Form(single("submit" -> text))
+
+  def handlePost(implicit ec: ExecutionContext, request: Request[_], formBinding: FormBinding): Future[Result] = {
+    emptyForm.bindFromRequest()
+      .fold(
+        formWithErrors => renderPage(formWithErrors),
+        formData => onSubmit(formData).flatMap { _ =>
+          val submitType = checkSubmitTypeForm.bindFromRequest().value.getOrElse("")
+          submitType match {
+            // The form was submitted using the 'continue' button. Execute the user-supplied logic.
+            case CONTINUE_BUTTON => onContinue(formData)
+            // The form was submitted using the 'filter' button. Save the filter settings and reload the page.
+            case FILTER_BUTTON =>
+              val searchFilter: SearchFilter = SearchAndFilterForm.form().bindFromRequest().get
+              for {
+                _ <- sessionCacheService.put(CLIENT_SEARCH_INPUT, searchFilter.search.getOrElse(""))
+                _ <- sessionCacheService.put(CLIENT_FILTER_INPUT, searchFilter.filter.getOrElse(""))
+                result <- Future.successful(Redirect(reloadCall(None, searchFilter.search, searchFilter.filter)))
+              } yield result
+            // The form was submitted using the 'clear filters' button. Clear the filters and reload the page.
```

Comment:
I'm not sure how I feel about this file, it's only for clients so far for `handlePost` - I get the principle of abstracting actions based on the submit value but it feels like the filtering should be separated out maybe? The controller is harder to read as well imo but we can discuss further